File: core/src/Visi/Expression.hs
module Visi.Expression ( 
                            Expression(LetExp, InnerLet, FuncExp, Apply,
                                       SinkExp, SourceExp, InvokeMethod,
                                       Var, BuiltIn, ValueConst,
                                       Group),
                            SourceLoc(NoSourceLoc, SourceFromURL, SourceLoc),
                           VarScope, LetScope, 
                           Type(TVar, TPrim, TOper, StructuralType),
                           FuncName(FuncName), 
                           LetId(LetId),
                           Prim(PrimDouble, PrimBool, PrimStr),
                           valuePrim,
                           SourcePoint, SourceSpan,
                           builtInLoc,
                           tFun, funcOperName,
                           defaultValueForType,
                           getSourceLoc,
                           HasSourceLoc,
                           SourceInfo,
                           Value(DoubleValue, StrValue, BoolValue, FuncValue, UndefinedValue)) where

{- ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is Visi.io.
 *
 * The Initial Developer of the Original Code is
 * David Pollak.
 * Portions created by the Initial Developer are Copyright (C) 2011-2012
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * ***** END LICENSE BLOCK ***** -}

import qualified Data.Map as Map
import qualified Data.Text as T
import Data.Char

newtype LetId = LetId T.Text deriving (Eq, Ord, Show)

type CanBeGeneric = Bool

type SourcePoint = (Int, Int)
type SourceSpan = (SourcePoint, SourcePoint)
type SourceInfo = (SourceSpan, T.Text)

data SourceLoc = NoSourceLoc 
                 | BuiltInSource String SourceLoc
                 | SourceFromURL String SourceInfo SourceLoc
                 | SourceLoc SourceInfo SourceLoc deriving (Show, Eq)

data Expression = LetExp SourceLoc LetId FuncName CanBeGeneric Type Expression
                  | InnerLet SourceLoc Type Expression Expression -- defines a Let plus an expression to be evaluated in the scope of the Let
                  | SinkExp SourceLoc LetId FuncName Type Expression
                  | SourceExp SourceLoc LetId FuncName Type
                  | InvokeMethod SourceLoc LetId FuncName Type
                  | FuncExp SourceLoc FuncName Type Expression
                  | Apply SourceLoc LetId Type Expression Expression
                  | Var SourceLoc FuncName
                  | BuiltIn SourceLoc FuncName Type (Value -> Value)
                  | ValueConst SourceLoc Value
                  | Group SourceLoc (Map.Map FuncName Expression) Type Expression

data Type = TVar T.Text
            | TPrim Prim
            | TOper T.Text [Type]
            | StructuralType (Map.Map T.Text Type) -- Structural type
            deriving (Eq, Ord)

newtype FuncName = FuncName T.Text deriving (Eq, Ord)

instance Show FuncName where
  show (FuncName name) = T.unpack name

-- | The name of the function operator
funcOperName = T.pack "->"

-- | create a type that represents a Function
tFun t1 t2 = TOper funcOperName [t1,t2]

startsWithLetter (a:b) = isLetter a


class HasSourceLoc a where
  getSourceLoc :: a -> SourceLoc

instance HasSourceLoc SourceLoc where
  getSourceLoc a = a

instance HasSourceLoc Expression where
  getSourceLoc (LetExp sl _ (FuncName name) _ t1 exp) = sl
  getSourceLoc (InnerLet sl _ letExp evalExp) = sl
  getSourceLoc (Apply sl _ t2 _ right) = sl
  getSourceLoc (FuncExp sl (FuncName param) rt exp) = sl
  getSourceLoc (InvokeMethod sl _ (FuncName name) _) = sl
  getSourceLoc (Var sl (FuncName name)) = sl
  getSourceLoc (ValueConst sl v) = sl
  getSourceLoc (Group sl map _ _) = sl
  getSourceLoc (BuiltIn sl (FuncName name) tpe _) = sl
  getSourceLoc (SinkExp sl _ (FuncName name) tpe _) = sl
  getSourceLoc (SourceExp sl _ (FuncName name) tpe) = sl

instance Show Expression where
  show (LetExp _ _ (FuncName name) _ t1 exp) = "let " ++ T.unpack name ++ " = " ++ show exp ++ " :: " ++ show t1
  show (InnerLet _ _ letExp evalExp) = show letExp ++ "\n" ++ show evalExp
  show (Apply _ _ t2 (Apply _ t1 _ (Var _ (FuncName name)) left) right) = show left ++ " " ++ T.unpack name ++ " " ++ show right
  show (FuncExp _ (FuncName param) rt exp) = "func " ++ T.unpack param ++ " = " ++ show exp ++ " :: " ++ show rt
  show (Apply _ _ _ e1 e2) = show e1 ++ "(" ++ show e2 ++ ")"
  show (InvokeMethod _ _ (FuncName name) _) = "#" ++ T.unpack name
  show (Var _ (FuncName name)) = T.unpack name
  show (ValueConst _ v) = show v
  show (Group _ map _ _) = "Group " ++ show map
  show (BuiltIn _ (FuncName name) tpe _) = T.unpack name ++ " :: " ++ show tpe
  show (SinkExp _ _ (FuncName name) tpe _) = "Sink: " ++ T.unpack name ++ " :: " ++ show tpe
  show (SourceExp _ _ (FuncName name) tpe) = "Source: " ++ T.unpack name ++ " :: " ++ show tpe



instance Show Type where
  show (TVar tv) = "TVar " ++ T.unpack tv
  show (TOper name (a:b:[])) | not $ startsWithLetter $ T.unpack name = show a ++ " " ++ T.unpack name ++ " " ++ show b
  show (TOper name params) = "TOper " ++ T.unpack name ++ ": " ++ show params
  show (TPrim prim) = show prim
  show (StructuralType info) = "StructuralType " ++ show info

data Prim = PrimDouble | PrimBool | PrimStr deriving (Eq, Ord)


defaultValueForType (TPrim PrimDouble) = DoubleValue 0.0
defaultValueForType (TPrim PrimStr) = StrValue $ T.pack ""
defaultValueForType (TPrim PrimBool) = BoolValue False
defaultValueForType _ = UndefinedValue

instance Show Prim where
  show PrimDouble = "Double"
  show PrimBool = "Bool"
  show PrimStr = "String"

valuePrim :: Value -> Maybe Type
valuePrim (DoubleValue _) = Just $ TPrim PrimDouble
valuePrim (StrValue _) = Just $ TPrim PrimStr
valuePrim (BoolValue _) = Just $ TPrim PrimBool
valuePrim _ = Nothing

builtInLoc name = BuiltInSource name NoSourceLoc

data Value = DoubleValue Double
             | StrValue T.Text
             | BoolValue Bool
             | UndefinedValue -- FIXME we hate undefined and it should go away 'cause it's null
             | FuncValue (Value -> Value)

instance Eq Value where
  DoubleValue d == DoubleValue d' = d == d'
  StrValue d == StrValue d' = d == d'
  BoolValue d == BoolValue d' = d == d'
  _ == _ = False

instance Show Value where
    show (DoubleValue i) = show i
    show (StrValue i) = show i
    show (BoolValue i) = show i
    show (FuncValue i) = "Value -> Value"



type VarScope = Map.Map FuncName Type

type LetScope = Map.Map FuncName Expression

File: core/scala/src/main/scala/visi/core/Expression.scala
package visi.core

/**
 * Created with IntelliJ IDEA.
 * User: dpp
 * Date: 12/5/12
 * Time: 1:49 PM
 * To change this template use File | Settings | File Templates.
 */
object Expression {
  type CanBeGeneric = Boolean

  type SourcePoint = Int
  type SourceSpan = (SourcePoint, SourcePoint)
  // type SourceInfo = (SourceSpan, String)

  type VarScope = Map[String, Type]

  type LetScope = Map[String, Expression]


  val FuncOperName = "->"

  def tFun(t1: Type, t2: Type): Type = TOper(FuncOperName, List(t1, t2))
}

import Expression._
import net.liftweb.util.Helpers

final case class LetId(id: String)
object LetId {
  def make: LetId = new LetId(Helpers.nextFuncName)
}

sealed trait Prim
case object PrimDouble extends Prim
case object PrimBool extends Prim
case object PrimStr extends Prim


object Type {
  def vendVar: Type = TVar(Helpers.nextFuncName)
}

sealed trait Type

final case class TVar(id: String) extends Type
final case class TPrim(prim: Prim) extends Type {
  override def toString = prim.toString
}
final case class TOper(opr: String, types: List[Type]) extends Type {
  override def toString = types match {
    case a :: b :: Nil if opr == Expression.FuncOperName => a.toString+" -> "+b.toString
    case _ => "TOper("+opr+" "+types.mkString(", ")+")"
  }
}
final case class StructuralType(structure: Map[String, Type]) extends Type

/**
 * The location of source
 */
sealed trait SourceLoc

final case object NoSourceLoc extends SourceLoc
final case class BuiltInSource(src: String, loc: SourceSpan) extends SourceLoc
final case class SourceFromURL(src: String, loc: SourceSpan) extends SourceLoc

/**
 * An expressionOrStruct
 */
sealed trait Expression {
  def loc: SourceLoc
  def tpe: Type
  def updateType(nt: Type): Expression
}

sealed trait HasName {
  def name: String
}

sealed trait HasLetId extends Expression {
  def id: LetId
}

final case class LetExp(loc: SourceLoc, id: LetId, name: String, generic: CanBeGeneric, tpe: Type, exp: Expression) extends Expression with HasLetId with HasName {
  override def toString = "Let "+name+" ("+tpe+") -> "+exp
  def updateType(nt: Type): Expression = this.copy(tpe = nt)
}

final case class InnerLet(loc: SourceLoc,
                           tpe: Type, exp1: Expression, exp2: Expression) extends Expression {
  def updateType(nt: Type): Expression = this.copy(tpe = nt)

}
final case class SinkExp(loc: SourceLoc,
                          id: LetId, name: String, tpe: Type, exp: Expression) extends Expression with HasLetId with HasName {
  def updateType(nt: Type): Expression = this.copy(tpe = nt)

}

final case class SourceExp(loc: SourceLoc,
                            id: LetId,
                            name: String,
                            tpe: Type) extends Expression with HasLetId with HasName {
  def updateType(nt: Type): Expression = this.copy(tpe = nt)
}

final case class FuncExp(loc: SourceLoc,
                         id: LetId,
                         name: String,
                         tpe: Type,
                         exp: Expression) extends Expression with HasLetId {
  override def toString = "Func "+name+" ("+tpe+") -> "+exp
  def updateType(nt: Type): Expression = this.copy(tpe = nt)
}

final case class Apply(loc: SourceLoc,
                        id: LetId,
                        tpe: Type,
                        exp1: Expression,
                        exp2: Expression) extends Expression with HasLetId {
  override def toString = "Apply("+tpe+" ["+exp1+"] ["+exp2+"])"
  def updateType(nt: Type): Expression = this.copy(tpe = nt)
}

final case class Var(loc: SourceLoc, id: LetId, name: String, tpe: Type) extends Expression with HasLetId {
  override def toString = "Var "+name+" "+tpe
  def updateType(nt: Type): Expression = this.copy(tpe = nt)
}
final case class BuiltIn(loc: SourceLoc, id: LetId, name: String, tpe: Type, func: StringBuilder => Unit) extends Expression with HasLetId with HasName {
  def updateType(nt: Type): Expression = this.copy(tpe = nt)

}
final case class ValueConst(loc: SourceLoc, value: Value, tpe: Type) extends Expression {
  def updateType(nt: Type): Expression = this.copy(tpe = nt)

}
final case class Group(map: Map[String, Expression]) extends Expression {
  def loc: SourceLoc = sys.error("No Sourceloc for Group")
  def tpe: Type = sys.error("No type for Group")
  def updateType(nt: Type): Expression = sys.error("No type for Group")
}


sealed trait Value {
  type T
  def value: T
  def toJsString: String
}
final case class DoubleValue(value: Double) extends Value{type T = Double

  def toJsString = value.toString
}
final case class StrValue(value: String) extends Value{type T = String
import Helpers._
  def toJsString = value.encJs

}
final case class BoolValue(value: Boolean) extends Value{type T = Boolean

  def toJsString = value.toString
}
//final case class FuncValue(value: Value => Value) extends Value{type T = Value => Value}
// UNDEFINED?

/**
 * The definition of a data structure
 *
 * @param loc the location in the source where the struct is defined
 * @param name the name of the data structure
 * @param typeParams the list of type parameters for this struct
 * @param fields the fields that are available on all sum subtypes of this struct
 * @param sums if the struct is a sum-type, then the sums and the fields describe the sums
 */
final case class Struct(loc: SourceLoc, name: String, typeParams: List[String], fields: List[StructField], sums: List[StructSum]) extends HasName

/**
 * Information about a field
 * @param name the name of the field
 * @param nominalType the nominal type of the field (this will be reified to an actual type in the typer)
 */
final case class StructField(name: String, nominalType: NominalType)

/**
 * The nominal type
 * @param name the name of the type (e.g., Number, String)
 */
final case class NominalType(name: String, params: List[String])

/**
 * A description of a sum type
 */
sealed trait StructSum {
  /**
   * The name of the sum element
   * @return the name of the sum element
   */
  def name: String
}

/**
 * A singleton (e.g., Empty)
 * @param name the name of the singleton
 */
final case class StructSingleton(name: String) extends StructSum

/**
 * A non-singleton sum type thingy (gotta get a better name here). Anyway, something like `Full(item)`
 *
 * @param name the name of the sum type thingy
 * @param fields the fields for the sum type thingy
 */
final case class StructSumItem(name: String, fields: List[StructField]) extends StructSum
